java/ql/src/semmle/code/java/dataflow/FlowSteps.qll

```diff
@@ -140,3 +140,63 @@ private class StringBuilderTaintPreservingCallable extends TaintPreservingCallab
     sink = -1
   }
 }
+
+private Type getAnExcludedParameterType() {
+  result instanceof PrimitiveType or
+  result.(RefType).hasQualifiedName("java.nio.charset", "Charset") or
+  result.(RefType).hasQualifiedName("java.util", "Locale")
+}
+
+private class SafeJoinArrayType extends PrimitiveType {
+  SafeJoinArrayType() { getName().regexpMatch("byte|double|float|int|long|short") }
+}
+
+private class ApacheStringUtilsTaintPreservingMethod extends TaintPreservingCallable {
```

Comment:
All of this might fit nicer in `java/ql/src/semmle/code/java/frameworks/apache/Lang.qll`. See also https://github.com/github/codeql/pull/5134